### ISSUE TITLE
[Bug]: Docker registry credential not found using `TESTCONTAINERS_HUB_IMAGE_NAME_PREFIX` (regression in `4.0.0`)

### ISSUE THREAD

**Issue by digital88** [testcontainers/testcontainers-dotnet#1286]
### Testcontainers version

4.0.0

### Using the latest Testcontainers version?

Yes

### Host OS

Linux

### Host arch

x86

### .NET version

8

### Docker version

[testcontainers.org 00:00:00.10] Connected to Docker:
  Host: tcp://docker:2375/
  Server Version: 27.3.1
  Kernel Version: 5.4.0-187-generic
  API Version: 1.47
  Operating System: Alpine Linux v3.20 (containerized)
  Total Memory: 15.62 GB

### Docker info

[testcontainers.org 00:00:00.10] Connected to Docker:
  Host: tcp://docker:2375/
  Server Version: 27.3.1
  Kernel Version: 5.4.0-187-generic
  API Version: 1.47
  Operating System: Alpine Linux v3.20 (containerized)
  Total Memory: 15.62 GB

### What happened?

Testcontainers stopped finding credentials for our custom docker registry in CI.

CI environment: gitlab CI, runners on k8s, we use dind as service to enable testcontainers in ci job.
We use custom docker registry on-prem (Harbor).

Credentials in CI stored in env var DOCKER_AUTH_CONFIG.

### Relevant log output

[testcontainers.org 00:00:00.10] Connected to Docker:
  Host: tcp://docker:2375/
  Server Version: 27.3.1
  Kernel Version: 5.4.0-187-generic
  API Version: 1.47
  Operating System: Alpine Linux v3.20 (containerized)
  Total Memory: 15.62 GB
[testcontainers.org 00:00:00.12] Docker config "/root/.docker/config.json" not found
[testcontainers.org 00:00:00.12] Searching Docker registry credential in Auths
[testcontainers.org 00:00:00.12] Docker registry credential our.selfhosted.registry/dockerhub not found
...
  Error Message:
   Docker.DotNet.DockerApiException : Docker API responded with status code=InternalServerError, response={"message":"unauthorized: unauthorized to access repository: dockerhub/postgres, action: pull: unauthorized to access repository: dockerhub/postgres, action: pull"}
  Stack Trace:
     at Docker.DotNet.DockerClient.HandleIfErrorResponseAsync(HttpStatusCode statusCode, HttpResponseMessage response)
...

When we did rollback to 3.8

[testcontainers.org 00:00:00.09] Connected to Docker:
  Host: tcp://docker:2375/
  Server Version: 27.3.1
  Kernel Version: 5.4.0-187-generic
  API Version: 1.47
  Operating System: Alpine Linux v3.20 (containerized)
  Total Memory: 15.62 GB
[testcontainers.org 00:00:00.11] Docker config "/root/.docker/config.json" not found
[testcontainers.org 00:00:00.12] Searching Docker registry credential in Auths
[testcontainers.org 00:00:00.12] Docker registry credential our.selfhosted.registry found
[testcontainers.org 00:00:10.00] Docker image our.selfhosted.registry/dockerhub/postgres:15.7 created
[testcontainers.org 00:00:19.59] Docker container bf12b593f26c created
...

### Additional information

Rollback to 3.8 made CI working again.
![Image](https://github.com/user-attachments/assets/d2e6361c-069c-4aea-ada1-bed12761782e)

**Comment by HofmeisterAn** [comment]
Oh, this is unfortunate. Could you please try removing `/dockerhub` from the `auths` node in the JSON value? It looks like we're resolving the hostname slightly differently, see `our.selfhosted.registry/dockerhub` versus `our.selfhosted.registry`. The `DOCKER_AUTH_CONFIG` should be something like this:

    {"auths":{"our.selfhosted.registry":{"auth":"bXlfdXNlcm5hbWU6bXlfcGFzc3dvcmQ="}}}

**Comment by digital88** [comment]
> Oh, this is unfortunate. Could you please try removing `/dockerhub` from the `auths` node in the JSON value? It looks like we're resolving the hostname slightly differently, see `our.selfhosted.registry/dockerhub` versus `our.selfhosted.registry`. The `DOCKER_AUTH_CONFIG` should be something like this:
> 
> ```
> {"auths":{"our.selfhosted.registry":{"auth":"bXlfdXNlcm5hbWU6bXlfcGFzc3dvcmQ="}}}
> ```

Our config is almost exactly the same as you posted:
`{"auths": {"https://our.selfhosted.registry": {"auth": "****"}}}`

`dockerhub` is the namespace or scope (not sure how it's called) where images stored. Consider: when you pull `postgres:latest` from dockerhub, you are pulling from `library/postgres` namespace/scope. Our selfhosted registry does the same, but with `dockerhub` if image doesn't have namespace, like `confluentinc/cp-kafka:7.7.1`. We actually just proxy dockerhub in our registry.

**Comment by HofmeisterAn** [comment]
I read it wrong. Try (I thought version 3.8.0 expected the "namespace", but version 4.0.0 does):

    {"auths":{"https://our.selfhosted.registry/dockerhub":{"auth":"bXlfdXNlcm5hbWU6bXlfcGFzc3dvcmQ="}}}

**Edit**: How do you pass the image to the container builder? Do you use `WithImage(IImage)`?

I ran a quick test, and at the moment, I don't really understand why you see the `/dockerhub` path in your log message. The following example uses the correct hostname to resolve the registry credentials.

```csharp
var dockerImage = new DockerImage("our.selfhosted.registry/dockerhub/postgres:15.7");
var registry = dockerImage.Registry;
var hostname = dockerImage.GetHostname();
Assert.Equal("our.selfhosted.registry", registry);
Assert.Equal("our.selfhosted.registry", hostname);
var container = new ContainerBuilder().WithImage(dockerImage).Build();
await container.StartAsync();
```

    [testcontainers.org 00:00:25.61] Docker config "C:\Users\andre.hofmeister\.docker\config.json" not found
    [testcontainers.org 00:00:25.61] Searching Docker registry credential in Auths
    [testcontainers.org 00:00:25.61] Docker registry credential our.selfhosted.registry found

**Comment by digital88** [comment]
> I read it wrong. Try (I thought version 3.8.0 expected the "namespace", but version 4.0.0 does):
> 
> ```
> {"auths":{"https://our.selfhosted.registry/dockerhub":{"auth":"bXlfdXNlcm5hbWU6bXlfcGFzc3dvcmQ="}}}
> ```
> 
> **Edit**: How do you pass the image to the container builder? Do you use `WithImage(IImage)`?

I will try to change auth later, thats not as simple, the variable is shared across different projects in Gitlab.

We pass image this way:
```
_postgresContainer = new PostgreSqlBuilder()
            .WithImage("postgres:15.7")
            .WithEnvironment(..
            .WithDatabase(..
            .WithUsername(..
            .WithPassword(..
            .WithResourceMapping(
                new FileInfo(Path.Combine(
                    CommonDirectoryPath.GetProjectDirectory().DirectoryPath,
                    "custom-pg-docker-entrypoint.sh")),
                new FileInfo("./custom-pg-docker-entrypoint.sh"),
                UnixFileModes.UserRead | UnixFileModes.GroupRead | UnixFileModes.UserExecute | UnixFileModes.GroupExecute)
            .WithEntrypoint("./custom-pg-docker-entrypoint.sh")
            .WithCommand("")
            .WithEnvironment(..
            .Build();
```

This is how our selfhosted harbor stores (caches) images:
![Image](https://github.com/user-attachments/assets/0d4712d3-d241-4eaf-ab61-8b0de2a9734a)

**Comment by HofmeisterAn** [comment]
Okay, I understand. You are using the custom configuration with the Docker Hub prefix. Now I can reproduce the issue. The following example fails as well:

```csharp
public class GitHub
{
    static GitHub()
    {
        TestcontainersSettings.HubImageNamePrefix = "our.selfhosted.registry/dockerhub";
    }

    [Fact]
    public async Task Issue1286()
    {
        var dockerImage = new DockerImage("postgres:15.7");
        var container = new ContainerBuilder().WithImage(dockerImage).Build();
        Assert.Equal("our.selfhosted.registry", container.Image.Registry);
        Assert.Equal("our.selfhosted.registry", container.Image.GetHostname());
        await container.StartAsync();
    }
}
```

    Assert.Equal() Failure: Strings differ
    Expected: "our.selfhosted.registry"
    Actual:   "our.selfhosted.registry/dockerhub"

Looks like something in [this](https://github.com/testcontainers/testcontainers-dotnet/blob/1a78654e9205d3ff01b807967a5408feec921622/src/Testcontainers/Builders/ContainerBuilder%603.cs#L83) line breaks it.

**Comment by digital88** [comment]
Good to see you was able to find root cause. Thanks.

**Comment by digital88** [comment]
> You are using the custom configuration with the Docker Hub prefix.

Yes, exactly. Sorry I forgot to mention. TESTCONTAINERS_HUB_IMAGE_NAME_PREFIX has this `our.selfhosted.registry/dockerhub`

**Comment by HofmeisterAn** [comment]
Okay, we need to extend the [`DockerImageFixture`](https://github.com/testcontainers/testcontainers-dotnet/blob/1a78654e9205d3ff01b807967a5408feec921622/tests/Testcontainers.Tests/Fixtures/Images/DockerImageFixture.cs) test data and consider the substitution of the registry:

```csharp
Add(new DockerImageFixtureSerializable(new DockerImage("postgres", null, null, null, "myregistry.azurecr.io")), "myregistry.azurecr.io/postgres", "myregistry.azurecr.io/postgres:latest");
Add(new DockerImageFixtureSerializable(new DockerImage("postgres", null, null, null, "myregistry.azurecr.io/library")), "myregistry.azurecr.io/library/postgres", "myregistry.azurecr.io/library/postgres:latest");
```

Basically, the overloaded [constructor](https://github.com/testcontainers/testcontainers-dotnet/blob/1a78654e9205d3ff01b807967a5408feec921622/src/Testcontainers/Images/DockerImage.cs#L107-L111) assigns the `/library` part to the `Registry` instead of to the `Repository` property:

|  | Actual | Expected |
|---|---|---|
| Registry | myregistry.azurecr.io/library | myregistry.azurecr.io |
| Repository | postgres | library/postgres |

_We need to be careful with the test [assertion](https://github.com/testcontainers/testcontainers-dotnet/blob/1a78654e9205d3ff01b807967a5408feec921622/tests/Testcontainers.Tests/Unit/Images/TestcontainersImageTest.cs#L53-L71). The error message is reversed compared to what is shown in the table, which is incorrect. The table is correct._

**Comment by digital88** [comment]
Version 3.10 also seems to work fine FWIW.

**Comment by HofmeisterAn** [comment]
The regression is due to the changes we made in the `IImage` interface (in version `4.0.0`). I have a potential fix and will create a PR later today.

**Comment by HofmeisterAn** [comment]
@digital88 #1287 should fix the issue. It would be great if you could validate the fix and check if it’s working in your environment.

**Comment by digital88** [comment]
> [@digital88](https://github.com/digital88) [#1287](https://github.com/testcontainers/testcontainers-dotnet/pull/1287) should fix the issue. It would be great if you could validate the fix and check if it’s working in your environment.

Thanks, I think I can try validate the fix. Is there a nupkg artifact that can be installed? Or I should wait for patch release?

**Comment by HofmeisterAn** [comment]
Thanks. I can publish a snapshot release by the end of this week. I would like to merge some changes, replacing the Docker client before.

**Comment by HofmeisterAn** [comment]
The fix is included in: `4.1.0-beta.11746011829`.

**Comment by digital88** [comment]
Can confirm `4.1.0-beta.11746011829` is working
